libminifi/src/c2/C2Client.cpp

```diff
@@ -0,0 +1,318 @@
+      std::string nameOption = option + ".name";
+
+      std::string name;
+      if (!configuration_->get(nameOption, name)) {
+        continue;
+      }
+      std::shared_ptr<state::response::ResponseNode> new_node = std::make_shared<state::response::ObjectNode>(name);
+      if (configuration_->get(classOption, class_definitions)) {
+        std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+        for (const std::string& clazz : classes) {
+          // instantiate the object
+          auto ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
+          if (nullptr == ptr) {
+            bool found_metric = [&] {
+              std::lock_guard<std::mutex> guard{metrics_mutex_};
+              auto metric = component_metrics_.find(clazz);
+              if (metric != component_metrics_.end()) {
+                ptr = metric->second;
+                return true;
+              }
+              return false;
+            }();
+            if (!found_metric) {
+              logger_->log_error("No metric defined for %s", clazz);
+              continue;
+            }
+          }
+          auto node = std::dynamic_pointer_cast<state::response::ResponseNode>(ptr);
+          std::static_pointer_cast<state::response::ObjectNode>(new_node)->add_node(node);
+        }
+
+      } else {
+        std::string optionName = option + "." + name;
+        auto node = loadC2ResponseConfiguration(optionName, new_node);
+      }
+
+      std::lock_guard<std::mutex> guard{metrics_mutex_};
+      root_response_nodes_[name] = new_node;
+    } catch (...) {
+      logger_->log_error("Could not create metrics class %s", metricsClass);
+    }
+  }
+}
+
+std::shared_ptr<state::response::ResponseNode> C2Client::loadC2ResponseConfiguration(const std::string &prefix, std::shared_ptr<state::response::ResponseNode> prev_node) {
+  std::string class_definitions;
+  if (!configuration_->get(prefix, class_definitions)) {
+    return prev_node;
+  }
+  std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+
+  for (const std::string& metricsClass : classes) {
+    try {
+      std::string option = prefix + "." + metricsClass;
+      std::string classOption = option + ".classes";
+      std::string nameOption = option + ".name";
+
+      std::string name;
+      if (!configuration_->get(nameOption, name)) {
+        continue;
+      }
+      std::shared_ptr<state::response::ResponseNode> new_node = std::make_shared<state::response::ObjectNode>(name);
+      if (name.find(',') != std::string::npos) {
+        std::vector<std::string> sub_classes = utils::StringUtils::split(name, ",");
+        for (const std::string& subClassStr : classes) {
+          auto node = loadC2ResponseConfiguration(subClassStr, prev_node);
+          if (node != nullptr)
+            std::static_pointer_cast<state::response::ObjectNode>(prev_node)->add_node(node);
+        }
+      } else {
```

Comment:
This feels like quite a convoluted logic with all these different branches. Do you feel like most of these branches are tested with our FlowControllerTests?

Reply:
I wouldn't bet on it, it has been moved from `FlowController.cpp`, and some branches already removed, this is the "improved" version (and it still looks like ****), as metrics currently provide us little to no utility (as far as I know), I don't think it is worth refactoring now